Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/src/TPCInterpolationSpec.cxx

```diff
@@ -37,37 +45,80 @@ void TPCInterpolationDPL::init(InitContext& ic)
   mTimer.Stop();
   mTimer.Reset();
   mInterpolation.init();
+  mResidualProcessor.init();
 }
 
 void TPCInterpolationDPL::run(ProcessingContext& pc)
 {
+  LOG(INFO) << "TPC Interpolation Workflow initialized. Start processing...";
   mTimer.Start(false);
-  const auto tracksITS = pc.inputs().get<gsl::span<o2::its::TrackITS>>("trackITS");
-  const auto tracksTPC = pc.inputs().get<gsl::span<o2::tpc::TrackTPC>>("trackTPC");
-  const auto tracksITSTPC = pc.inputs().get<gsl::span<o2::dataformats::TrackTPCITS>>("match");
-  const auto tracksTPCClRefs = pc.inputs().get<gsl::span<o2::tpc::TPCClRefElem>>("trackTPCClRefs");
-  const auto trackMatchesTOF = pc.inputs().get<gsl::span<o2::dataformats::MatchInfoTOF>>("matchTOF"); // FIXME missing reader
-  const auto clustersTOF = pc.inputs().get<gsl::span<o2::tof::Cluster>>("clustersTOF");
-
-  const auto& inputsTPCclusters = getWorkflowTPCInput(pc);
-
-  // pass input data to TrackInterpolation object
-  mInterpolation.setITSTracksInp(tracksITS);
-  mInterpolation.setTPCTracksInp(tracksTPC);
-  mInterpolation.setTPCTrackClusIdxInp(tracksTPCClRefs);
-  mInterpolation.setTPCClustersInp(&inputsTPCclusters->clusterIndex);
-  mInterpolation.setTOFMatchesInp(trackMatchesTOF);
-  mInterpolation.setITSTPCTrackMatchesInp(tracksITSTPC);
-  mInterpolation.setTOFClustersInp(clustersTOF);
+  RecoContainer recoData;
+  recoData.collectData(pc, *mDataRequest.get());
+
+  std::vector<o2::globaltracking::RecoContainer::GlobalIDSet> gidTables;
+  std::vector<o2::track::TrackParCov> seeds;
+  std::vector<float> trkTimes;
+  std::vector<GTrackID> gids;
+  bool processITSTPConly = mProcessITSTPConly; // so that the flag can be used inside the lambda
+  // the creator goes from most complete track (ITS-TPC-TRD-TOF) to least complete one (ITS-TPC)
+  auto creator = [&gidTables, &seeds, &trkTimes, &recoData, &processITSTPConly, &gids](auto& _tr, GTrackID _origID, float t0, float tErr) {
+    if constexpr (std::is_base_of_v<o2::track::TrackParCov, std::decay_t<decltype(_tr)>>) {
+      bool trackGood = true;
+      bool hasOuterPoint = false;
+      auto gidTable = recoData.getSingleDetectorRefs(_origID);
+      if (!gidTable[GTrackID::ITS].isIndexSet() || !gidTable[GTrackID::TPC].isIndexSet()) {
```

Comment:
This will make the creator to skip the AfterBurner tracks, since instead of ITS they have ITSAB reference, but this should be ok, I would not use the tracks with only a couple of ITS hits for the constraint.

Reply:
I think it would be good to not exclude the ITSAB tracks here, since a few lines below I cut on the minimum number of ITS clusters anyways. So I have changed that now.

Reply:
@martenole The ITSAB may have any number of clusters in 2-7 range (can be limited by options), so, rejection base on Nclusters only will not work. And since there is separate ITS track associated to ITSAB, the further code using ITS track outer param will not work. 
In any case, most probably we will not run the afterburner in the sync. mode, I would simply skip these tracks.

Reply:
Ok I see, thanks. I will change this back again then.